numpy/lib/_nanfunctions_impl.py

```diff
@@ -1646,14 +1646,20 @@ def _nanquantile_ureduce_func(
         wgt = None if weights is None else weights.ravel()
         result = _nanquantile_1d(part, q, overwrite_input, method, weights=wgt)
     else:
+        #GH-25731
+        axis = _nx.normalize_axis_tuple(axis, a.ndim)[0]
         result = np.apply_along_axis(_nanquantile_1d, axis, a, q,
                                      overwrite_input, method, weights)
         # apply_along_axis fills in collapsed axis with results.
         # Move that axis to the beginning to match percentile's
         # convention.
         if q.ndim != 0:
-            result = np.moveaxis(result, axis, 0)
-
+            #GH-25731
+            num_dims = len(a.shape)+len(q.shape)-1
+            new_axes = list(range(len(q.shape), num_dims))
```

Comment:
```suggestion
            new_axes = list(range(0, q.ndim))
```
Shouldn't this do the trick? We want the dimensions of q to be the first ones.

Reply:
@lorentzenchr I am pretty sure I am misunderstanding your proposed solution. However just changing `new_axes` in the place you proposed will result in a value error `ValueError: repeated axis in `destination` argument` in `np.moveaxis()`.  (As a concrete example I consider, `a = np.ones((3, 4))`
`q = np.full((1, 2), 0.5)` and `np.nanquantile(a, q, axis=1).shape)`)